### PR TITLE
Reduce the number of NSQ blocks based on simple motion

### PR DESCRIPTION
**Pull request by NaderMahdi**
Reduce the number of nsq blocks tested in MD based on the cost and variance of the predicted square block using its simple motion vector from motion estimation. When it is ON it leads to 0.2% BDrate loss for 5% speed up in context of M0 encoder mode on the full AOM clip set.

### CODE REVIEW

**Comment by hassount** [comment]
@NaderMahdi 
what is the memory footprint impact of this ?

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 1]
```diff
@@ -1411,6 +1411,19 @@ EbErrorType signal_derivation_multi_processes_oq(
         // GM_DOWN                                    Downsampled resolution with a downsampling factor of 2 in each dimension
         // GM_TRAN_ONLY                               Translation only using ME MV.
         picture_control_set_ptr->gm_level = GM_FULL;
+#endif
+#if SIMPLE_MOTION_SEARCH_SPLIT
+        // Perform simple motion search before none_partition to decide if we
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: Please clarify the meaning of "Perform simple motion search before none_partition".

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 2]
```diff
@@ -1411,6 +1411,19 @@ EbErrorType signal_derivation_multi_processes_oq(
...
+        // want to remove all partitions other than PARTITION_SPLIT. If set to 0, this
+        // model is disabled. If set to 1, the model attempts to perform
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: If "model" means "feature", please change it to feature in the comments.

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c [hunk 3]
```diff
@@ -1411,6 +1411,19 @@ EbErrorType signal_derivation_multi_processes_oq(
...
+        // PARTITION_SPLIT only. If set to 2, the model also attempts to prune
+        // PARTITION_SPLIT.
+        picture_control_set_ptr->simple_motion_search_split = picture_control_set_ptr->slice_type == I_SLICE ? 0 : 0;
+        // Added weight for the used threshold to be tuned for different encoder modes
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: Consider instead "Weight to be used in tuning the threshold for the different encoder modes"

File: Source/Lib/Common/Codec/EbPictureControlSet.h [hunk 4]
```diff
@@ -14338,6 +14338,13 @@ extern "C" {
 #endif
 #if GM_OPT
         uint8_t                                gm_level;
+#endif
+#if SIMPLE_MOTION_SEARCH_SPLIT
+        // The aggresiveness of pruning with simple_motion_search.
+        // Currently 0 is the lowest, and 2 the highest.
+        int simple_motion_search_prune_agg;
+        int simple_motion_search_split;
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbPictureControlSet.h]: Add a comment to indicate what this flag indicates.

File: Source/Lib/Common/Codec/EbCodingUnit.h [hunk 5]
```diff
@@ -583,7 +583,23 @@ extern "C" {
...
     } SuperBlock;
-
+#if SIMPLE_MOTION_SEARCH_SPLIT
+typedef struct PC_TREE {
+  struct PC_TREE *split[4];
+  // Simple motion search_features
+  unsigned int sms_none_feat[2];
+  unsigned int sms_rect_feat[8];
+  int sms_none_valid;
+  int sms_rect_valid;
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbCodingUnit.h]: Add comments to explain the meaning of all the variables in the structure.

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c (hunk 1 above)
Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: Done

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c (hunk 2 above)
Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: Done

File: Source/Lib/Common/Codec/EbPictureDecisionProcess.c (hunk 3 above)
Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbPictureDecisionProcess.c]: Done

File: Source/Lib/Common/Codec/EbPictureControlSet.h (hunk 4 above)
Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbPictureControlSet.h]: Done

File: Source/Lib/Common/Codec/EbCodingUnit.h (hunk 5 above)
Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbCodingUnit.h]: Done

**Comment by hassount** [comment]
Closing this PR as the performance is not an M0 level

**Pull request closed without merging**